File: frontend/lib/crypto/safety_numbers.dart
import 'dart:typed_data';
import 'package:cryptography/cryptography.dart' as cg;
import 'wire_format.dart';

// Pairwise safety number for MITM detection

//   SafetyNumber(A, B) =
//     first 30 decimal digits of :
//       SHA256(SORT_LEXICOGRAPHIC(IK_pub_A, IK_pub_B) || album_id)
//     formatted as: XXXXX XXXXX XXXXX XXXXX XXXXX XXXXX

// IKs are the Ed25519 32-byte identity keys. The sort makes the function
// symmetric : both A and B compute the same number for a given album
class SafetyNumber {
  static const int totalDigits = 30;
  static const int groupSize = 5;

  // Returns the unformatted 30 digit string (no spaces)
  static Future<String> compute({
    required Uint8List ikPubA,
    required Uint8List ikPubB,
    required Uint8List albumId,
  }) async {
    if (ikPubA.length != kIkPubLen) {
      throw ArgumentError(
          'ikPubA must be $kIkPubLen bytes, got ${ikPubA.length}');
    }
    if (ikPubB.length != kIkPubLen) {
      throw ArgumentError(
          'ikPubB must be $kIkPubLen bytes, got ${ikPubB.length}');
    }
    if (albumId.length != kAlbumIdLen) {
      throw ArgumentError(
          'albumId must be $kAlbumIdLen bytes, got ${albumId.length}');
    }

    final lo = _lex(ikPubA, ikPubB) <= 0 ? ikPubA : ikPubB;
    final hi = _lex(ikPubA, ikPubB) <= 0 ? ikPubB : ikPubA;

    final preimage = Uint8List(kIkPubLen + kIkPubLen + kAlbumIdLen);
    preimage.setRange(0, kIkPubLen, lo);
    preimage.setRange(kIkPubLen, 2 * kIkPubLen, hi);
    preimage.setRange(2 * kIkPubLen, preimage.length, albumId);

    final hash = await cg.Sha256().hash(preimage);
    return _firstNDecimalDigits(Uint8List.fromList(hash.bytes), totalDigits);
  }

  // Returns the formatted "XXXXX XXXXX XXXXX XXXXX XXXXX XXXXX"
  static Future<String> formatted({
    required Uint8List ikPubA,
    required Uint8List ikPubB,
    required Uint8List albumId,
  }) async {
    final digits =
        await compute(ikPubA: ikPubA, ikPubB: ikPubB, albumId: albumId);
    return _group(digits);
  }

  static int _lex(Uint8List a, Uint8List b) {
    final n = a.length < b.length ? a.length : b.length;
    for (var i = 0; i < n; i++) {
      if (a[i] != b[i]) return a[i] - b[i];
    }
    return a.length - b.length;
  }

  // SHA-256 (32 bytes, BE) is < 10^78. Format as a 78-char zero-padded decimal
  // string and slice the first n. This makes "first 30 decimal digits" a fixed
  // positional contract regardless of whether the hash's leading byte is zero
  static String _firstNDecimalDigits(Uint8List digest, int n) {
    // Big-endian → BigInt
    BigInt v = BigInt.zero;
    for (final b in digest) {
      v = (v << 8) | BigInt.from(b);
    }
    const int totalLen = 78; // 2^256 < 10^78
    final s = v.toString().padLeft(totalLen, '0');
    return s.substring(0, n);
  }

  static String _group(String digits) {
    final groups = <String>[];
    for (var i = 0; i < digits.length; i += groupSize) {
      groups.add(digits.substring(
          i, i + groupSize > digits.length ? digits.length : i + groupSize));
    }
    return groups.join(' ');
  }
}

File: server/internal/crypto/safety_numbers.go
package crypto

import (
	"crypto/sha256"
	"fmt"
	"math/big"
	"strings"
)

// SafetyNumber computes the pairwise MITM fingerprint :

//	first 30 decimal digits of SHA256(sort_lex(IK_pub_A, IK_pub_B) || album_id)

// The lex sort makes this symmetric (A,B) == (B,A). Used for user visible
// out-of-band verification, so byte parity with Dart matters

// Returns the 30-digit unformatted string. Use SafetyNumberFormatted for the
// space grouped display form
const safetyNumberDigits = 30

// 2^256 < 10^78, so any SHA-256 fits in 78 decimal digits when zero-padded
// slicing the leading 30 makes the digits a fixed positional contract :
// hashes that happen to start with low bytes never produce shorter strings
const safetyNumberPad = 78

func SafetyNumber(ikPubA, ikPubB, albumID []byte) (string, error) {
	if len(ikPubA) != IkPubLen {
		return "", fmt.Errorf("crypto: ikPubA must be %d bytes, got %d", IkPubLen, len(ikPubA))
	}
	if len(ikPubB) != IkPubLen {
		return "", fmt.Errorf("crypto: ikPubB must be %d bytes, got %d", IkPubLen, len(ikPubB))
	}
	if len(albumID) != AlbumIDLen {
		return "", fmt.Errorf("crypto: albumID must be %d bytes, got %d", AlbumIDLen, len(albumID))
	}
	lo, hi := ikPubA, ikPubB
	if compareBytes(ikPubA, ikPubB) > 0 {
		lo, hi = ikPubB, ikPubA
	}
	pre := make([]byte, 0, IkPubLen+IkPubLen+AlbumIDLen)
	pre = append(pre, lo...)
	pre = append(pre, hi...)
	pre = append(pre, albumID...)
	d := sha256.Sum256(pre)

	n := new(big.Int).SetBytes(d[:]).String()
	if len(n) < safetyNumberPad {
		n = strings.Repeat("0", safetyNumberPad-len(n)) + n
	}
	return n[:safetyNumberDigits], nil
}

// SafetyNumberFormatted returns the 30 digits in six space separated groups
// of five : "XXXXX XXXXX XXXXX XXXXX XXXXX XXXXX"
func SafetyNumberFormatted(ikPubA, ikPubB, albumID []byte) (string, error) {
	d, err := SafetyNumber(ikPubA, ikPubB, albumID)
	if err != nil {
		return "", err
	}
	var sb strings.Builder
	for i := 0; i < len(d); i += 5 {
		if i > 0 {
			sb.WriteByte(' ')
		}
		sb.WriteString(d[i : i+5])
	}
	return sb.String(), nil
}

func compareBytes(a, b []byte) int {
	n := len(a)
	if len(b) < n {
		n = len(b)
	}
	for i := 0; i < n; i++ {
		if a[i] != b[i] {
			return int(a[i]) - int(b[i])
		}
	}
	return len(a) - len(b)
}
